src/main/kotlin/com/jetbrains/snakecharm/codeInsight/SnakemakeParamsInShellReferenceContributor.kt

```diff
@@ -5,20 +5,30 @@ import com.intellij.patterns.PlatformPatterns
 import com.intellij.psi.*
 import com.intellij.psi.util.PsiTreeUtil
 import com.intellij.util.ProcessingContext
+import com.jetbrains.python.psi.PyCallExpression
 import com.jetbrains.python.psi.PyStringLiteralExpression
 import com.jetbrains.snakecharm.codeInsight.completion.SMKKeywordCompletionContributor
 import com.jetbrains.snakecharm.lang.SnakemakeNames
 import com.jetbrains.snakecharm.lang.psi.SMKParamsReference
 import com.jetbrains.snakecharm.lang.psi.SMKRuleParameterListStatement
+import com.jetbrains.snakecharm.lang.psi.SMKRuleRunParameter
 import java.util.regex.Pattern
 
 class SnakemakeParamsInShellReferenceContributor : PsiReferenceContributor() {
+    private val insideRuleSection = PlatformPatterns.psiElement()
+            .inFile(SMKKeywordCompletionContributor.IN_SNAKEMAKE)
+            .inside(SMKRuleParameterListStatement::class.java)
+    private val insideCallExpressionInRuleRunParameter = PlatformPatterns.psiElement()
+            .inFile(SMKKeywordCompletionContributor.IN_SNAKEMAKE)
+            .inside(PyCallExpression::class.java)
+            .withAncestor(6, PlatformPatterns.psiElement(SMKRuleRunParameter::class.java))
```

Comment:
we could easily break such low-level checks, e.g. place "shell" call inside one of several `if` statements. 

NB: Is always please do not use low level impl details of PSI / AST tree outside of PsiElements. PSI and AST could be changed in future or by language injections or etc, it shouldn't break all features at once due to hardcoded things like this.